### DISCUSSION TITLE
Stopping a stream with Langchain LlamCpp in python

### DISCUSSION THREAD

**Discussion by calypset** [ggml-org/llama.cpp#6212] [Q&A]
Hi, I am looking to stop a stream that is ongoing for any given reason. I set up a stream with the handler as follows, I have a queue and a thread that manages downstream. I have been trying a few things but so far unsucessful. I figured I could pass a stop signal as a token but unsure how. 

The callback class is: 
```
from langchain.callbacks.base import BaseCallbackHandler

class QueueCallback(BaseCallbackHandler):
    def __init__(self, q):
        self.q = q

    def cancel(self, q):
        return self.q.empty()

    def on_llm_new_token(self, token: str, **kwargs: Any) -> None:
        self.q.put(token)

    def on_llm_end(self, *args, **kwargs: Any) -> None:
        return self.q.empty()
```

And model loaded: 
```
from langchain_community.llms import LlamaCpp

llm = LlamaCpp(
        model_path=model_source,
        **params,
        stop=['</s>', '###'], 
        callbacks=[QueueCallback(q)]
)
```
Anyone has an idea on how to achieve such hard stream stop seamlessly? 

PS: Llama.cpp is fantastic!

**Comment by reddiamond1234** [comment]
I am catching a Keyboard Interrupt exception (ctrl + c)...

**Reply by calypset** [reply]
I guess my question is how to tell the Llm to stop streaming the answer

**Reply by reddiamond1234** [reply]
well there isn't

**Reply by calypset** [reply]
So, then I guess it is true that once a stream from LlamaCpp has started, there is no way to stop it, and the program needs to wait until the stream ends before invoking the chain once again

**Reply by reddiamond1234** [reply]
As far as I know

**Comment by SirNate0** [comment]
I believe this feature has been added, I think you just need to close the http connection: https://github.com/ggml-org/llama.cpp/issues/9273
